### PR TITLE
BkWrap #399

### PR DESCRIPTION
**Pull request by HDouss**
#399 : Added BkWrap class. BkParallel, BkTimeable, BkSafe and BkReuse extend the newly created BkWrap.

### CODE REVIEW

**Comment by davvd** [comment]
@hdouss Many thanks for the PR, let me find a reviewer for it

**Comment by davvd** [comment]
@dmzaytsev please review, thanks

**Comment by dmzaytsev** [comment]
@HDouss please update the pull request description

according our [quality rules](http://at.teamed.io/qa.html)
> Pull request description explains the solution proposed and contains a link to the original ticket it is related to.

File: src/main/java/org/takes/http/BkParallel.java [hunk 1]
```diff
@@ -63,44 +55,44 @@ public BkParallel(final Back back) {
...
      * @param threads Threads total
      */
     public BkParallel(final Back back, final int threads) {
-        this(
-            back,
-            Executors.newFixedThreadPool(
-                threads, new BkParallel.Threads()
-            )
+        this(back,
+                Executors.newFixedThreadPool(
```
Comment (dmzaytsev) [inline src/main/java/org/takes/http/BkParallel.java]: @HDouss let's use 4 spaces indentation
I wonder why Quilce doesn't complain here

File: src/main/java/org/takes/http/BkParallel.java [hunk 2]
```diff
@@ -63,44 +55,44 @@ public BkParallel(final Back back) {
...
-                    } catch (final IOException ex) {
-                        throw new IllegalStateException(ex);
-                    }
-                }
+        super(new Back() {
+            @Override
+            @SuppressWarnings("PMD.DoNotUseThreads")
+            public void accept(final Socket socket) {
+                svc.execute(
+                        new Runnable() {
```
Comment (dmzaytsev) [inline src/main/java/org/takes/http/BkParallel.java]: @HDouss let's use 4 spaces indentation here and below

File: src/main/java/org/takes/http/BkParallel.java [hunk 3]
```diff
@@ -111,16 +103,16 @@ public void run() {
...
             final Thread thread = new Thread(runnable);
             thread.setName(
-                String.format(
-                    "%s-%d",
-                    BkParallel.class.getSimpleName(),
-                    this.total.getAndAdd(1)
-                )
+                    String.format(
+                            "%s-%d",
+                    BkParallel.class.getSimpleName(), this.total.getAndAdd(1)
```
Comment (dmzaytsev) [inline src/main/java/org/takes/http/BkParallel.java]: @HDouss third argument should be on next line
like this
```
String.format(
    "%s-%d",
    BkParallel.class.getSimpleName(), 
    this.total.getAndAdd(1)
)
```

File: src/main/java/org/takes/http/BkWrap.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import lombok.EqualsAndHashCode;
+
+/**
+ * Back Wrap.
+ * @author Hamdi Douss (douss.hamdi@gmail.com)
+ * @version $Id$
+ *
+ */
+@EqualsAndHashCode(of = "origin")
+public class BkWrap implements Back {
```
Comment (dmzaytsev) [inline src/main/java/org/takes/http/BkWrap.java]: @HDouss what about a unit test for this class?

**Comment by dmzaytsev** [comment]
@HDouss please see a few comments above

**Comment by HDouss** [comment]
@dmzaytsev Comments took into consideration. For the unit test, I did not add one. What should the unit test actually test ? I think there is no nedd for a unit test here, like for RqWrap, TkWrap and RsWrap.

**Comment by dmzaytsev** [comment]
@HDouss in my opinion this unit test should check that BkWrap calls #accept() of origin Back and socket passed into the method

**Comment by HDouss** [comment]
@dmzaytsev This is already tested in the BkParallel, BkTimeable, BkSafe and BkReuse tests. Testing BkWrap would rather test the inheritance. If you still convinced it has to have a unit test, I will make a puzzle for it.

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@elenavolokhova please, check this issue for QA compliance, as per [par.24](http://at.teamed.io/policy.html#24)

**Comment by davvd** [comment]
@rultor deploy

**Comment by elenavolokhova** [comment]
@dmzaytsev 
According to our [quality rules](http://www.teamed.io/qa.html):
 > Comments were mostly about design problems, not cosmetic issues.
 
 There were mostly cosmetic issues found during review. Try to pay attention to bigger design problems and the overall quality of the code.
Please confirm that this rule is clear.

**Comment by dmzaytsev** [comment]
@elenavolokhova I confirm

**Comment by elenavolokhova** [comment]
@davvd Quality is acceptable here.

**Comment by davvd** [comment]
> @davvd Quality is acceptable here.

@elenavolokhova thanks a lot, next time everybody should try to make it better

**Comment by davvd** [comment]
@dmzaytsev added 10 mins to @elenavolokhova, for QA review, payment code is `71812775`; Much obliged! I have added **24 mins** to your account in payment "71812778", 23 hours and 22 mins spent; review comments (c=9) added as a bonus; +24 to your rating, your total score is [+953](http://www.netbout.com/b/35166?open=rating)

**Pull request merged**